src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
-                return emittedWhileDraining;
-            }
-            return 0;
-        }
-
-        final Func1<InnerSubscriber<T>, Boolean> DRAIN_ACTION = new Func1<InnerSubscriber<T>, Boolean>() {
-
-            @Override
-            public Boolean call(InnerSubscriber<T> s) {
-                if (s.q != null) {
-                    long r = mergeProducer.requested;
-                    int emitted = s.drainQueue();
-                    if (emitted > 0) {
-                        s.requestMore(emitted);
+                } finally {
+                    if (!skipFinal) {
+                        synchronized (this) {
+                            emitting = false;
+                        }
                     }
-                    if (emitted == r) {
-                        // we emitted as many as were requested so stop the forEach loop
-                        return Boolean.FALSE;
+                }
+                // in case there was some activity, continue with the regular emission loop
+                emitLoop();
+            } else {
+                // either someone else is emitting or there was no available request
+                // queue up the scalar
+                RxRingBuffer svq = scalarQueue;
+                if (svq == null) {
+                    svq = RxRingBuffer.getSpscInstance();
+                    add(svq);
+                    scalarQueue = svq;
+                }
+                try {
+                    svq.onNext(nl.next(value));
+                } catch (MissingBackpressureException ex) {
+                    onError(ex);
+                    return;
+                } catch (IllegalStateException ex) {
+                    // if there was an IAE and we are unsubscribed, just quit
+                    if (!isUnsubscribed()) {
+                        onError(ex);
                     }
+                    return;
                 }
-                return Boolean.TRUE;
-            }
-
-        };
-
-        @Override
-        public void onError(Throwable e) {
-            if (!completed) {
-                completed = true;
-                innerError(e, true);
+                // then try to emit it (or notify the running emission loop).
+                emit();
             }
         }
         
-        private void innerError(Throwable e, boolean parent) {
-            if (delayErrors) {
-                synchronized (this) {
-                    if (exceptions == null) {
-                        exceptions = new ConcurrentLinkedQueue<Throwable>();
+        void fastEmitNext(T value) {
+            synchronized (this) {
+                if (emitting) {
+                    int s = fastSize;
+                    Object[] o = fastArray;
+                    if (o == null) {
+                        o = new Object[4];
+                        fastArray = o;
+                    } else
+                    if (s == o.length) {
+                        Object[] o2 = new Object[s + (s >> 2)];
```

Comment:
This is capable of unbounded queue growth and breaks backpressure.

Here is a simple test:

```java
        Observable.merge(
                Observable.range(0, 10000000).subscribeOn(Schedulers.computation()), 
                Observable.range(10000001, 20000000)
            )
        .subscribe(new Subscriber<Integer>() {

            @Override
            public void onCompleted() {
            }

            @Override
            public void onError(Throwable e) {
            }

            @Override
            public void onNext(Integer t) {
                try {
                    Thread.sleep(100); // simulate computation time
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                System.out.println(t);
            }
            
        });
```

If I put this code in here after the resize occurs:

```java
System.out.println("resize list = " + o.length);
```

Then I get this output:

```
resize list = 5
resize list = 6
resize list = 7
resize list = 8
resize list = 10
resize list = 12
resize list = 15
resize list = 18
resize list = 22
resize list = 27
resize list = 33
resize list = 41
resize list = 51
resize list = 63
resize list = 78
resize list = 97
resize list = 121
resize list = 151
resize list = 188
resize list = 235
resize list = 293
resize list = 366
resize list = 457
resize list = 571
resize list = 713
resize list = 891
resize list = 1113
resize list = 1391
resize list = 1738
resize list = 2172
resize list = 2715
resize list = 3393
resize list = 4241
resize list = 5301
resize list = 6626
resize list = 8282
resize list = 10352
resize list = 12940
resize list = 16175
resize list = 20218
resize list = 25272
resize list = 31590
resize list = 39487
resize list = 49358
resize list = 61697
resize list = 77121
resize list = 96401
resize list = 120501
resize list = 150626
resize list = 188282
resize list = 235352
resize list = 294190
resize list = 367737
resize list = 459671
resize list = 574588
resize list = 718235
resize list = 897793
resize list = 1122241
resize list = 1402801
```